pkg/registry/core/bastion/storage/tableconvertor.go

```diff
@@ -0,0 +1,116 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/duration"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+var swaggerMetadataDescriptions = metav1.ObjectMeta{}.SwaggerDoc()
+
+type convertor struct {
+	headers []metav1beta1.TableColumnDefinition
+}
+
+func newTableConvertor() rest.TableConvertor {
+	return &convertor{
+		headers: []metav1beta1.TableColumnDefinition{
+			{Name: "Name", Type: "string", Format: "name", Description: swaggerMetadataDescriptions["name"]},
+			{Name: "Shoot", Type: "string", Format: "name", Description: "The Shoot this bastion belongs to."},
+			{Name: "Seed", Type: "string", Format: "name", Description: "The Seed cluster on which the Shoot is scheduled."},
+			{Name: "IP", Type: "string", Format: "name", Description: "The bastion host's IP and/or hostname."},
+			{Name: "Heartbeat", Type: "string", Format: "name", Description: "The time and date when the last heartbeat occured."},
+			{Name: "Expires", Type: "string", Format: "name", Description: "The time and date after which the bastion will be deleted."},
+			{Name: "Age", Type: "date", Description: swaggerMetadataDescriptions["creationTimestamp"]},
+		},
+	}
+}
+
+// ConvertToTable converts the output to a table.
+func (c *convertor) ConvertToTable(ctx context.Context, obj runtime.Object, tableOptions runtime.Object) (*metav1beta1.Table, error) {
+	var (
+		err   error
+		table = &metav1beta1.Table{
+			ColumnDefinitions: c.headers,
+		}
+	)
+
+	if m, err := meta.ListAccessor(obj); err == nil {
+		table.ResourceVersion = m.GetResourceVersion()
+		table.SelfLink = m.GetSelfLink()
+		table.Continue = m.GetContinue()
+	} else {
+		if m, err := meta.CommonAccessor(obj); err == nil {
+			table.ResourceVersion = m.GetResourceVersion()
+			table.SelfLink = m.GetSelfLink()
+		}
+	}
+
+	table.Rows, err = metatable.MetaToTableRow(obj, func(obj runtime.Object, m metav1.Object, name, age string) ([]interface{}, error) {
+		var (
+			bastion = obj.(*core.Bastion)
+			cells   = []interface{}{}
+		)
+
+		ingress := bastion.Status.Ingress
+
+		cells = append(cells, bastion.Name)
+		cells = append(cells, bastion.Spec.ShootRef.Name)
+
+		if bastion.Spec.SeedName == nil {
+			cells = append(cells, "<pending>")
+		} else {
+			cells = append(cells, *bastion.Spec.SeedName)
+		}
+
+		if ingress == nil || (ingress.IP == "" && ingress.Hostname == "") {
+			cells = append(cells, "<pending>")
+		} else if ingress.Hostname != "" {
+			cells = append(cells, ingress.Hostname)
+		} else {
+			cells = append(cells, ingress.IP)
+		}
+
+		lastHeartbeat := "<never>"
+		if !bastion.Status.LastHeartbeatTimestamp.IsZero() {
+			lastHeartbeat = fmt.Sprintf("%s ago", duration.HumanDuration(time.Since(bastion.Status.LastHeartbeatTimestamp.Time)))
+		}
+		cells = append(cells, lastHeartbeat)
+
+		expires := "<pending>"
+		if !bastion.Status.ExpirationTimestamp.IsZero() {
+			expires = metatable.ConvertToHumanReadableDateType(*bastion.Status.ExpirationTimestamp)
+		}
```

Comment:
```go
	LastHeartbeatTimestamp *metav1.Time
	ExpirationTimestamp *metav1.Time
```

possible nil pointer dereference?
```go
		if !bastion.Status.LastHeartbeatTimestamp.IsZero() {
 // ..
		if !bastion.Status.ExpirationTimestamp.IsZero() {
```

Reply:
`IsZero` is safe for nil values: https://github.com/kubernetes/apimachinery/blob/v0.21.0/pkg/apis/meta/v1/time.go#L60-L62